File: PHP/LAURA/interwikis.php
<?php
  /* Ce fichier comporte les fonctions utilisées pour établir les liens interwikis entre Vikidia francophone,
     Vikidia hispanophone et WikiKids (réciproquement), et de ces trois wikis en direction de Grundschulwiki,
     Simple English Wikipedia et Wikipédia dans sa version linguistique adaptée (francophone, hispanophone ou
     néerlandophone) unilatéralement. Ces fonctions sont écrites de manière abstraite et peuvent donc être
     généralisées pour servir avec n’importe quels interwikis, à condition que la fonction requeteAPI() du
     fichier base.php connaisse l’adresse des API correspondantes.

   Ce code est écrit par thilp et placé sous licence GPL 3.
  */

require_once "edit.php";

  // Variable globale de type array indexé comprenant les préfixes des interlangues pris en charge :
$GLOBALE_tabInterlangues = array('es','fr','nl');
  // Variable globale de type array indexé comprenant les préfixes des liens unilatéraux pris en charge :
$GLOBALE_tabLiensUnilateraux = array('wp','de','en');

function prefixe2nom($prefixe,$langue_site='fr'){ // Retourne la traduction du nom du wiki correspondant à $prefixe dans la langue $langue_site.
  switch($prefixe){
    case 'fr':
      switch($langue_site){
        case 'es': return 'Vikidia en francés';
        case 'nl': return 'franse Vikidia';
        default: return 'French Vikidia';
      }
    case 'wp':
      switch($langue_site){
        case 'fr': return 'Wikipédia';
        default: 'Wikipedia';
      }
    case 'nl': return 'Wikikids';
    case 'es':
      switch($langue_site){
        case 'fr': return 'Vikidia hispanphone';
        case 'nl': return 'spaanse Vikidia';
        default: 'Spanish Vikidia';
      }
    case 'en': return 'Simple English Wikipedia';
    case 'de': return 'Grundschulwiki';
    default:
      switch($langue_site){
        case 'fr': return '(préfixe wiki inconnu)';
        case 'nl': return '(wiki prefix onbekend)';
        case 'es': return '(wiki prefijo desconocidos)';
        default: return '(unknown wiki prefix)';
      }
  }
}

function article_existe($titre,$prefixe){ // Teste l’existence d’un article intitulé $titre dans le wiki correspondant à $prefixe.
  $retour = requeteAPI('action=query&prop=revisions&titles='.$titre.'&rvprop=content&format=xml',$prefixe);
  return strpos($retour,' missing=""') === false;
}

function est_a_vikifier_econome($texte,$titre_prefixe_vers,$prefixe_de,$prefixe_vers){ // Teste l’opportunité d’interwikifier un article et renvoie le titre traduit si oui, false sinon.
  return stripos($texte,'[['.$prefixe_vers.':') === false && article_existe($titre_prefixe_vers,$prefixe_vers);
}

function resume_publication_interwiki_multiples($prefixe_de,$tabInterwikis){ // Renvoie le résumé adapté pour la publication d’un interwiki vers $prefixe_vers sur le wiki correspondant à $prefixe_de.
  $chaine = '';
  foreach($tabInterwikis as $clef => $valeur){
    $chaine.= ', '.prefixe2nom($clef,$prefixe_de);
  }
  $chaine = substr($chaine,1).'.';
  switch($prefixe_de){
    case 'fr': return 'Liens interwiki automatiques vers'.$chaine;
    case 'es': return 'Automáticas interwiki enlaces con dirección a'.$chaine;
    case 'nl': return 'Automatische interwiki links naar'.$chaine;
    default: return 'Automatic interwiki links towards'.$chaine;
  }
}

function traduction_titre($titre,$tabPrefixes){ // Renvoie un tableau associant chaque préfixe de $tabPrefixes à la traduction de $titre en la langue correspondant au préfixe.
  $tabRetour = array('fr' => $titre);
  $chaine_er_prefixes = '';
  foreach($tabPrefixes as $prefixe){
    if($prefixe != 'fr'){ $chaine_er_prefixes = $chaine_er_prefixes.'|'.$prefixe; }
  }
  $chaine_er_prefixes = substr($chaine_er_prefixes,1);
  preg_match_all('#\[\[('.$chaine_er_prefixes.'):([^\]]+)\]\]#',lectureArticle_wp($titre),$tabMatches,PREG_SET_ORDER);
  foreach($tabMatches as $serie){
    $pref = $serie[1];
    $nom = $serie[2];
    if(resolution_existence_page_et_redirection($nom,$pref,$redirection,$page_redirect)){
      if($redirection){
        $tabRetour[$pref] = $page_redirect; }
      else{ $tabRetour[$pref] = $nom; }
    }
  }
  $tabRetour['wp'] = $titre;
    // Gestion des titres non trouvés avec fr.wikipédia
  $tabInterwikis_manquants = array_diff_key(array_flip($tabPrefixes),$tabRetour);
  foreach($tabInterwikis_manquants as $interwiki => $truc){
    $trad_wikt = traduction_titre_wiktionary($titre,$interwiki);
    if($trad_wikt !== ''){
      $tabRetour[$interwiki] = $trad_wikt;
    }
  }
  return $tabRetour;
}

function traduction_titre_wiktionary($titre,$prefixe_vers,$prefixe_de='fr'){ // Tente de traduire $titre (mot de la langue correspondant à $prefixe_de) dans la langue correspondant à $prefixe_vers avec les Wiktionnaires.
    // Tentative dans le Wiktionnaire francophone.
  $retour = cURL('http://fr.wiktionary.org/w/api.php','','action=query&prop=revisions&rvprop=content&format=xml&titles='.urlencode(strtolower($titre)));
  if(preg_match('#\{\{trad[+\-]*\|'.$prefixe_vers.'\|([^}]+)\}\}#U',$retour,$tabMatches)){
    return $tabMatches[1]; }
  else{ // Tentative dans le Wiktionnaire $prefixe_vers-phone.
    $retour = cURL('http://'.$prefixe_vers.'.wiktionary.org/w/api.php','','action=query&prop=revisions&rvprop=content&format=xml&titles='.urlencode(strtolower($titre)));
    switch($prefixe_vers){
      case 'de':
        if(preg_match('#\{\{'.$prefixe_vers.'\}\}: \[1\] \[\[([^\]]+)\]\]#',$retour,$tabMatches)){
          return $tabMatches[1]; }
      case 'es':
        if(preg_match('#;1: ?\[\[([^|]+)\|#',$retour,$tabMatches)){
          return $tabMatches[1]; }
      case 'nl':
        if(preg_match("~{{=${prefixe_de}[^=]*=}}.*'''$titre'''[^#]*#[^[]*\[\[([^\]]+)\]\]~iU",$retour,$tabMatches)){
          return $tabMatches[1]; }
      default: return false;
    }
  }
}

function interwikification_totale_lexico($titre){
  global $GLOBALE_tabInterlangues,$GLOBALE_tabLiensUnilateraux;
  $tabInterwikis = array_merge($GLOBALE_tabInterlangues,$GLOBALE_tabLiensUnilateraux);
    // Traduction de $titre dans les langues de $tabInterwikis.
  $tabTraductions_titre = traduction_titre($titre,$tabInterwikis);
  foreach($GLOBALE_tabInterlangues as $prefixe_de){ // Pour chaque wiki concerné par le robot...
    $titre = $tabTraductions_titre[$prefixe_de];
    $texte = lectureArticle($titre,$prefixe_de);
    $tabInterwikis_a_faire = array();
    foreach($tabInterwikis as $prefixe_vers){ // Création des interwikis nécessaires.
      if($prefixe_de != $prefixe_vers && isset($tabTraductions_titre[$prefixe_vers]) && est_a_vikifier_econome($texte,$tabTraductions_titre[$prefixe_vers],$prefixe_de,$prefixe_vers)){
        $tabInterwikis_a_faire[$prefixe_vers] = '[['.$prefixe_vers.':'.$tabTraductions_titre[$prefixe_vers].']]';
      }
    }
      // Tri des interwikis selon la méthode Paulochon : WP d’abord, ordre lexicographique pour les autres.
    ksort($tabInterwikis_a_faire,SORT_STRING);
    if(array_key_exists('wp',$tabInterwikis_a_faire)){
      $tabInterwikis_a_faire = array_merge(array('wp' => '[[wp:'.$titre.']]'),array_diff_key($tabInterwikis_a_faire,array('wp'=>true)));
    }
      // Édition de l’article.
    $chaine = '';
    foreach($tabInterwikis_a_faire as $interwiki){
      $chaine.= "\n$interwiki";
    }
    echo "\n$titre : $chaine\n";
    ajoutTexte($titre,$chaine,resume_publication_interwiki_multiples($prefixe_de,$tabInterwikis_a_faire),'bot',$prefixe_de);
  }
}

interwikification_totale_lexico('Chien');
?>

File: Perl/v3/interwikis.pl
#! /usr/bin/perl -w

require "base.pl";
require "edit.pl";

##
##########
#######################
####################################################
##############  CONFIGURATION HERE #################
####################################################

$local_language = 'fr';

####################################################
########### NO MORE CONFIGURATION NOW ##############
####################################################
########################
#########
##

# Prefixes with the two-sided interwiki links
@tab_bilateral = ('es','fr','nl');
# Prefixes with the one-sided interwiki links
@tab_unilateral = ('wp','de','en');

# Return the translation of the wiki name linked to $prefix in the
# $sitelang language
sub name_of_prefix
{
  my ($prefix, $sitelang) = @_;
  $sitelang = $local_language if (!defined($sitelang));

  if ($prefix eq 'fr')
  {
    return 'Vikidia en francés' if ($sitelang eq 'es');
    return 'franse Vikidia' if ($sitelang eq 'nl');
    return 'French Vikidia';
  }
  elsif ($prefix =~ /^wp_/)
  {
    return 'Wikipédia' if ($sitelang eq 'fr');
    return 'Wikipedia';
  }
  elsif ($prefix eq 'nl')
  {
    return 'WikiKids';
  }
  elsif ($prefix eq 'es')
  {
    return 'Vikidia hispanophone' if ($sitelang eq 'fr');
    return 'spaanse Vikidia' if ($sitelang eq 'nl');
    return 'Spanish Vikidia';
  }
  elsif ($prefix eq 'en')
  {
    return 'Simple English Wikipedia';
  }
  elsif ($prefix eq 'de')
  {
    return 'Grundschulwiki';
  }
  else
  {
    return '(préfixe wiki inconnu)' if ($sitelang eq 'fr');
    return '(wiki prefix onbekend)' if ($sitelang eq 'nl');
    return '(wiki prefijo desconocidos)' if ($sitelang eq 'es');
    return '(unknown wiki prefix)';
  }
}

# Checks if the $title article exists in the wiki that matches $prefix
sub article_exists
{
  my ($title, $prefix) = @_;
  return (requeteAPI('action#=query#&prop#=revisions#&titles#='.$title.
      '#&rvprop#=content', $prefix) !~ / missing=""/);
}

# Checks if the article $text has to be interwikified
sub has_to_be_interwikified
{
  my ($text, $title_prefix_to, $prefix_from, $prefix_to) = @_;
  return ($text !~ /\[\[$prefix_to:/ and
    article_exists($title_prefix_to, $prefix_to));
}

# Give the summary corresponding to the publication of an interwiki link
# to the linked wikis on the $prefix_from wiki
sub summary
{
  my ($prefix_from, $ref_htab_interwikis) = @_;
  my $wikis = '';

  foreach (keys(%$ref_htab_interwikis))
  {
    $wikis .= ', '.name_of_prefix($_, $prefix_from);
  }
  $wikis =~ s/^.//;
  $wikis .= '.';

  return 'Liens interwikis automatiques vers'.$wikis if ($prefix_from eq 'fr');
  return 'Automáticas interwiki enlaces con dirección a'.$wikis
    if ($prefix_from == 'es');
  return 'Automatische interwiki links naar'.$wikis if ($prefix_from eq 'nl');
  return 'Automatic interwiki links towards'.$wikis;
}

# Uses the Wiktionaries to translate $title from the $prefix_from
# language to the $prefix_to one.
sub translate_with_wikt
{
  my ($title, $prefix_to, $prefix_from) = @_;
  $prefix_from = 'fr' if (!defined($prefix_from));
  $title =~ tr/A-Z/a-z/;

  # Trying in the French Wiktionary
  return $1 if (requeteAPI('action#=query#&prop#=revisions#&revprop#=content'.
      '#&titles#='.$title, 'wikt_fr') =~
      /\{\{trad[+-]*\|$prefix_to\|([^}]+)\}\}/);
  # Trying in the $prefix_to's Wiktionary
  my $res = requeteAPI('action#=query#&prop#=revisions#&revprop#=content'.
      '#&titles#='.$title, 'wikt_'.$prefix_to);
  return $1 if ($prefix_to eq 'de' and $res =~
    /\{\{$prefix_to\}\}: \[1\] \[\[([^\]]+)\]\]/);
  return $1 if ($prefix_to eq 'es' and $res =~ /;1: ?\[\[([^\]])\|/);
  return $1 if ($prefix_to eq 'nl' and $res =~
    /\{\{=$prefixe_from[^=]*=\}\}.*'''$title'''[^#]*#[^[]*\[\[([^\]]+)\]\]/i;
  return undef;
}

# Returns a hashtable A => B where A is a prefix in $tab_prefixes and B
# is the translation of $title in the language corresponding to the prefix,
# using the language-specific Wikipedia
sub translate_title
{
  my $title = shift;
  my @tab_prefixes = @_;
  my $htab_translate[$local_language] = $title;
  my $htab_translate['wp'] = $title;
  my $wp_article = lectureArticle($title, 'wp_'.$local_language);

  foreach (@tab_prefixes)
  {
    my ($translated) = ($wp_article =~ /\[\[$_:([^\]]+)\]\]/)
      if ($_ ne $local_language);
    if (defined($translated))
    {
      my ($bool_redirect, $page_redirect) = (0);
      if (exists_or_redirects($translated, $_, $bool_redirect, $page_redirect))
      {
	$htab_translate[$_] = $page_redirect if ($bool_redirect);
	$htab_translate[$_] = $translated if (!$bool_redirect);
      }
    }
    else
    { # Management of the not-found-on-wp titles using Wiktionary
      my $trad_wikt = translate_with_wikt($title, $prefix);
      $htab_translate[$_] = $trad_wikt;
    }
  }
  return %htab_translate;
}
